### PR TITLE
Updating Iterators to be compatible with C++17 in MSVC

### PR DESCRIPTION
**Pull request by Kwizatz**
The std::iterator class is being deprecated on MSVC++,
and currently if the compilation flag /std:c++latest
is used a warning is issued in this regard if any
iterators use the class as a base class.

If an external source file being compiled includes
the repeated_field.h header, the iterator clases
RepeatedPtrIterator and RepeatedPtrOverPtrsIterator
trigger the warning.

This change solves the warning and should avoid it in
the future when the default is to deprecate the class.

### CODE REVIEW

**Comment by Kwizatz** [comment]
Here is the full text of of the warnings this change addresses:

c:\projects\aeonengine\vcpkg\installed\x64-windows\include\google\protobuf\repeated_field.h(2165): warning C4996: 'std::iterator<std::random_access_iterator_tag,Element,ptrdiff_t,_Ty*,_Ty&>': warning STL4015: The std::iterator class template (used as a base class to provide typedefs) is deprecated in C++17. (The <iterator> header is NOT deprecated.) The C++ Standard has never required user-defined iterators to derive from std::iterator. To fix this warning, stop deriving from std::iterator and start providing publicly accessible typedefs named iterator_category, value_type, difference_type, pointer, and reference. Note that value_type is required to be non-const, even for constant iterators. You can define _SILENCE_CXX17_ITERATOR_BASE_CLASS_DEPRECATION_WARNING or _SILENCE_ALL_CXX17_DEPRECATION_WARNINGS to acknowledge that you have received this warning. [C:\projects\aeonengine\build\engine\ProtoBufClasses.vcxproj]
  c:\projects\aeonengine\vcpkg\installed\x64-windows\include\google\protobuf\repeated_field.h(2160): note: see declaration of 'std::iterator<std::random_access_iterator_tag,Element,ptrdiff_t,_Ty*,_Ty&>'
  c:\projects\aeonengine\vcpkg\installed\x64-windows\include\google\protobuf\repeated_field.h(2245): note: see reference to class template instantiation 'google::protobuf::internal::RepeatedPtrIterator<Element>' being compiled
c:\projects\aeonengine\vcpkg\installed\x64-windows\include\google\protobuf\repeated_field.h(2259): warning C4996: 'std::iterator<std::random_access_iterator_tag,Element,ptrdiff_t,_Ty*,_Ty&>': warning STL4015: The std::iterator class template (used as a base class to provide typedefs) is deprecated in C++17. (The <iterator> header is NOT deprecated.) The C++ Standard has never required user-defined iterators to derive from std::iterator. To fix this warning, stop deriving from std::iterator and start providing publicly accessible typedefs named iterator_category, value_type, difference_type, pointer, and reference. Note that value_type is required to be non-const, even for constant iterators. You can define _SILENCE_CXX17_ITERATOR_BASE_CLASS_DEPRECATION_WARNING or _SILENCE_ALL_CXX17_DEPRECATION_WARNINGS to acknowledge that you have received this warning. [C:\projects\aeonengine\build\engine\ProtoBufClasses.vcxproj]
  c:\projects\aeonengine\vcpkg\installed\x64-windows\include\google\protobuf\repeated_field.h(2160): note: see declaration of 'std::iterator<std::random_access_iterator_tag,Element,ptrdiff_t,_Ty*,_Ty&>'
  c:\projects\aeonengine\vcpkg\installed\x64-windows\include\google\protobuf\repeated_field.h(2328): note: see reference to class template instantiation 'google::protobuf::internal::RepeatedPtrOverPtrsIterator<Element,VoidPtr>' being compiled

File: src/google/protobuf/repeated_field.h [hunk 1]
```diff
@@ -2301,21 +2292,14 @@ class RepeatedPtrIterator
...
-class RepeatedPtrOverPtrsIterator
-    : public std::iterator<std::random_access_iterator_tag, Element> {
+class RepeatedPtrOverPtrsIterator {
  public:
   typedef RepeatedPtrOverPtrsIterator<Element, VoidPtr> iterator;
-  typedef std::iterator<std::random_access_iterator_tag, Element> superclass;
-
-  // Shadow the value_type in std::iterator<> because const_iterator::value_type
-  // needs to be T, not const T.
-  typedef typename std::remove_const<Element>::type value_type;
```
Comment (BillyONeal) [inline src/google/protobuf/repeated_field.h]: You accidentally broke this remove_const. From the error message: "Note that value_type is required to be non-const, even for constant iterators."

Comment (Kwizatz) [reply src/google/protobuf/repeated_field.h]: This is true for the other iterator as well, thanks for the heads up, I'll upload a new commit soon.

**Comment by jeffreylindsey** [comment]
This fixes issue #4539.

Personally, I'd appreciate this being merged as soon as possible, as it fixes the last issue preventing code using protobuf from being cleanly built with C++17.

**Comment by Kwizatz** [comment]
> Josh could you also take a look. I think this is fine.

Please @joshkel @jhump or @haberman, take a look, this isn't such a big change, nor critical, but having to keep maintaining a fork up to date just for this is getting ridiculous 😢.

**Comment by haberman** [comment]
Sorry I missed Gerben's latest message. LGTM.

**Comment by Kwizatz** [comment]
@haberman No worries, thanks! 😄👍

**Comment by TeBoring** [comment]
Broken internal tests. Revert this for now. Need to land this from internal first.

**Comment by Kwizatz** [comment]
@TeBoring Anything I could do to help with this? let me know.

**Comment by R0ll1ngSt0ne** [comment]
Any plans to finally release this? It would be nice to get rid of all these deprecation warnings...

**Comment by haberman** [comment]
I think this has been released already.  In 3.10 (at least) `RepeatedPtrIterator` no longer derives from std::iterator: https://github.com/protocolbuffers/protobuf/blob/3.10.x/src/google/protobuf/repeated_field.h#L2254

However it appears that there are other iterators that are still deriving from std::iterator, like `RepeatedFieldBackInsertIterator`, `RepeatedPtrFieldBackInsertIterator`, `AllocatedRepeatedPtrFieldBackInsertIterator`, and `UnsafeArenaAllocatedRepeatedPtrFieldBackInsertIterator`. Possibly those are what is causing the warnings for you now?

**Pull request merged**